### PR TITLE
Fix issues for server side rendering in Next.js

### LINKED ISSUE #11219: Errors rendering Editor in Next.js
## Bug Description

I was trying to create Story Editor in Next.js however there were several errors that appear one after the other ( Most of them are due to Next.js SSR )

image

( Gets fixed by removing [CSS import ](https://github.com/GoogleForCreators/web-stories-wp/blob/afe02f9209112c2447559dbfa797e3192b052f96/packages/story-editor/src/components/form/dateTime/datePicker.js#L20) )
Reference: https://nextjs.org/docs/messages/css-npm

image

( Gets fixed by using `globalThis` )

image

( Gets fixed by removing `"type": "module"` from `package.json` )

image

image

image

image

image

image

## Expected Behaviour

Editor and dashboard should work without any issues in NextJs.

## Steps to Reproduce

Step - 1: `npx create-next-app@latest`

Step - 2: Use snippet from [getting-started](https://github.com/GoogleForCreators/web-stories-wp/blob/main/docs/third-party-integration/story-editor/getting-started.md#step-3-use-the-storyeditor-component-to-render-the-editor) to use the editor

## Additional Context

- Plugin Version:
- WordPress Version:
- Operating System:
- Browser:

### PR DESCRIPTION
**Pull request by mohdsayed**
## Context

See #11219

## Testing Instructions

- [x] This is a non-user-facing change and requires no QA

This PR can be tested by following these steps:

In web-stories-wp
1. Bundle and publish all packages locally.

In Next.js setup
1. `npx create-next-app@latest`
2. `npm i react@17 react-dom@17` as next.js is using v18
3. Use snippet from [getting-started](https://github.com/GoogleForCreators/web-stories-wp/blob/main/docs/third-party-integration/story-editor/getting-started.md#step-3-use-the-storyeditor-component-to-render-the-editor) to use the editor and dashboard
4. Install all the local dependencies ( or the ones which are updated in this PR )
```bash
npm --registry http://localhost:4873/ install --save \
@googleforcreators/media \
@googleforcreators/animation \
@googleforcreators/dashboard \
@googleforcreators/date \
@googleforcreators/design-system \
@googleforcreators/dom \
@googleforcreators/element-library \
@googleforcreators/fonts \
@googleforcreators/i18n \
@googleforcreators/masks \
@googleforcreators/migration \
@googleforcreators/moveable \
@googleforcreators/output \
@googleforcreators/react \
@googleforcreators/rich-text \
@googleforcreators/stickers \
@googleforcreators/story-editor \
@googleforcreators/templates \
@googleforcreators/text-sets \
@googleforcreators/tracking \
@googleforcreators/units \
@googleforcreators/url
@googleforcreators/elements
```
6. `npm run dev`

## Checklist

- [ ] This PR addresses an existing issue and I have linked this PR to it in ZenHub
- [x] I have tested this code to the best of my abilities
- [ ] I have verified accessibility to the best of my abilities ([docs](https://github.com/googleforcreators/web-stories-wp/blob/main/docs/accessibility-testing.md))
- [ ] I have verified i18n and l10n (translation, right-to-left layout) to the best of my abilities
- [ ] This code is covered by automated tests (unit, integration, and/or e2e) to verify it works as intended ([docs](https://github.com/googleforcreators/web-stories-wp/tree/main/docs#testing))
- [x] I have added documentation where necessary
- [x] I have added a matching `Type: XYZ` label to the PR

---

Fixes #11219

### CODE REVIEW

File: packages/story-editor/src/components/tabview/index.js [hunk 1]
```diff
@@ -317,7 +317,12 @@ TabView.propTypes = {
   tabs: PropTypes.array.isRequired,
   tab: PropTypes.string.isRequired,
   tabRefs: PropTypes.objectOf(
-    PropTypes.shape({ current: PropTypes.instanceOf(Element) })
+    PropTypes.shape({
+      current:
+        typeof Element !== 'undefined'
+          ? PropTypes.instanceOf(Element)
+          : PropTypes.any,
+    }) // To handle SSR
```
Comment (swissspidy) [inline packages/story-editor/src/components/tabview/index.js]: Does `PropTypes.elementType` work instead?

File: packages/react/src/useFocusOut.js [hunk 2]
```diff
@@ -53,7 +53,6 @@ function useFocusOut(ref, callback, deps) {
       node.removeEventListener('focusout', onFocusOut);
       doc.removeEventListener('pointerdown', onDocumentClick, opts);
     };
-    // eslint-disable-next-line react-hooks/exhaustive-deps -- Pass through provided deps.
```
Comment (swissspidy) [inline packages/react/src/useFocusOut.js]: We should update `additionalHooks` in the ESLint config so we can keep this

Comment (mohdsayed) [reply packages/react/src/useFocusOut.js]: I didn't remove this line, not sure why eslint removes it automatically

image

Comment (mohdsayed) [reply packages/react/src/useFocusOut.js]: I think this is why

image

Comment (swissspidy) [reply packages/react/src/useFocusOut.js]: After updating the `additionalHooks` config this should not happen anymore and we can add it back

**Comment by sayedtaqui** [comment]
With these many updates, the editor and dashboard both are working in Next.js 🥳  .

File: packages/dashboard/src/utils/usePagePreviewSize.js [hunk 3]
```diff
@@ -97,26 +97,28 @@ const sizeFromWidth = (
...
     ? MIN_DASHBOARD_WIDTH
     : windowWidth;
 };
 
 export default function usePagePreviewSize(options = {}) {
   const { thumbnailMode = false, isGrid } = options;
+  const container =
+    typeof document !== 'undefined' ? document.getElementById(WPBODY_ID) : null;
   // When the dashboard is pulled out of wordpress this id will need to be updated.
   // For now, we need to grab wordpress instead because of how the app's rendered
```
Comment (swissspidy) [inline packages/dashboard/src/utils/usePagePreviewSize.js]: Sounds like something we should refactor now and move to `wp-dashboard` somehow...

File: packages/dashboard/src/utils/keyboardOnlyOutline.js [hunk 4]
```diff
@@ -55,8 +55,11 @@ const KeyboardOnlyOutline = () => {
     }
   };
 
-  document.addEventListener('keydown', handleKeydown);
-  document.addEventListener('mousedown', handleMousedown);
+  // @todo Should this be in useEffect?
```
Comment (swissspidy) [inline packages/dashboard/src/utils/keyboardOnlyOutline.js]: 👍

File: packages/story-editor/src/components/tabview/index.js (hunk 1 above)
Comment (mohdsayed) [reply packages/story-editor/src/components/tabview/index.js]: It doesn't because `PropTypes.elementType` is for React Element.

image

File: packages/react/src/useFocusOut.js (hunk 2 above)
Comment (mohdsayed) [reply packages/react/src/useFocusOut.js]: Oh I see, updated in https://github.com/GoogleForCreators/web-stories-wp/pull/11270/commits/843e32a8465dc8b8643c68345c9d9f44aa524bc7

File: packages/dashboard/src/utils/keyboardOnlyOutline.js (hunk 4 above)
Comment (mohdsayed) [reply packages/dashboard/src/utils/keyboardOnlyOutline.js]: Done in #11275

File: packages/dashboard/src/utils/usePagePreviewSize.js (hunk 3 above)
Comment (mohdsayed) [reply packages/dashboard/src/utils/usePagePreviewSize.js]: Missed out!  #11276

File: .eslintrc [hunk 5]
```diff
@@ -685,6 +691,30 @@
...
         ]
       }
+    },
+    {
+      "files": [
+        "packages/story-editor/src/**/*.js",
+        "packages/dashboard/src/**/*.js",
+        "packages/design-system/src/**/*.js",
+        "packages/element-library/src/**/*.js"
+      ],
```
Comment (swissspidy) [inline .eslintrc]: This probably affects a bunch more packages.

Might be easiest to do `"files": [ "packages/**/*.js" ],` and then add all non-editor packages (e.g. widget, lightbox, etc.) to `excludedFiles`

File: packages/dashboard/src/app/router/route.js [hunk 6]
```diff
@@ -63,7 +63,7 @@ export function resolveRoute(route) {
      * If the url has no root append it to the current route and create
      * a nested root within the context of the dashboard app.
      */
-    return `${window.location.hash}/${route}`;
+    return `${globalThis.location?.hash}/${route}`;
```
Comment (swissspidy) [inline packages/dashboard/src/app/router/route.js]: On the server `globalThis.location?.hash` is `undefined`, so this will return `undefined/${route}`

Could mitigate this using something like

```suggestion
    const hash = globalThis.location?.hash || '';
    return `${hash}/${route}`;
```

File: packages/dashboard/src/app/router/routerProvider.js [hunk 7]
```diff
@@ -83,8 +85,8 @@ function RouterProvider({ children, ...props }) {
         defaultRoute,
       },
       actions: {
-        push: history.current.push,
-        replace: history.current.replace,
+        push: history.current?.push,
+        replace: history.current?.replace,
```
Comment (swissspidy) [inline packages/dashboard/src/app/router/routerProvider.js]: Should we return `noop`s here instead of possibly `null`?

File: packages/dashboard/src/app/router/routerProvider.js [hunk 8]
```diff
@@ -63,7 +65,7 @@ function RouterProvider({ children, ...props }) {
   };
 
   const [queryParams, setQueryParams] = useState(
-    parse(history.current.location.search)
+    parse(history.current?.location?.search)
```
Comment (swissspidy) [inline packages/dashboard/src/app/router/routerProvider.js:69]: If `history.current?.location?.search` is `null` then this will be `parse(null)` which will return `{ null: '' }`, which seems unexpected and like something that could possible cause bugs.

Might be worth adding JS unit tests with

```js
/**
 * @jest-environment node
 */
```

to test this provider

File: packages/dashboard/src/app/router/routerProvider.js [hunk 9]
```diff
@@ -45,9 +45,11 @@ export const getActiveRoute = ({
   return matchingRoutes.length <= 1 ? defaultRoute : currentPath;
 };
 function RouterProvider({ children, ...props }) {
-  const history = useRef(props.history || createHashHistory());
+  const history = useRef(
+    props.history || (globalThis.document && createHashHistory())
+  ); // createHashHistory internally uses document.
   const [currentPath, setCurrentPath] = useState(
-    history.current.location.pathname
+    history.current?.location?.pathname
```
Comment (swissspidy) [inline packages/dashboard/src/app/router/routerProvider.js]: Perhaps

```suggestion
    history.current?.location?.pathname || ''
```

File: packages/design-system/src/components/keyboard/index.js [hunk 10]
```diff
@@ -59,7 +59,8 @@ const globalRef = createRef();
 
 function setGlobalRef() {
   if (!globalRef.current) {
-    globalRef.current = document.documentElement;
+    globalRef.current =
+      typeof document !== 'undefined' ? document.documentElement : null;
```
Comment (swissspidy) [inline packages/design-system/src/components/keyboard/index.js]: Why not `globalThis.document?.documentElement || null`

Generally speaking, seems like this PR uses both different ways (`globalThis.document?.` vs `typeof document !== 'undefined'`). Can we settle on one?

File: packages/dom/src/stripHTML.js [hunk 11]
```diff
@@ -14,7 +14,8 @@
  * limitations under the License.
  */
 
-const buffer = document.createElement('div');
+const buffer =
+  typeof document !== 'undefined' ? document.createElement('div') : {};
 
 export default function stripHTML(string) {
```
Comment (swissspidy) [inline packages/dom/src/stripHTML.js:20]: Do we need a `stripHTML` function that actually works on the server?

File: packages/element-library/src/video/onDropHandler.js [hunk 12]
```diff
@@ -17,7 +17,10 @@
 function onDropHandler(dropTargetId) {
   // Auto-play video on drop
   setTimeout(() => {
-    const videoEl = document.getElementById(`video-${dropTargetId}`);
+    const videoEl =
+      typeof document !== 'undefined'
+        ? document.getElementById(`video-${dropTargetId}`)
+        : null;
```
Comment (swissspidy) [inline packages/element-library/src/video/onDropHandler.js]: Is the change here needed? This function should never be called on the server

File: packages/i18n/src/translateWithMarkup.js [hunk 13]
```diff
@@ -120,8 +120,10 @@ function TranslateWithMarkup({ mapping = {}, children }) {
     );
   }
 
-  const node = new DOMParser().parseFromString(children, 'text/html').body
-    .firstChild;
+  const node =
+    typeof window !== 'undefined'
+      ? new DOMParser().parseFromString(children, 'text/html').body.firstChild
```
Comment (swissspidy) [inline packages/i18n/src/translateWithMarkup.js]: Ah, forgot about that.

Does this change break anything in `TranslateWithMarkup`?

Wonder if some Jest unit tests with `@jest-environment node` would be useful here as well.

File: packages/story-editor/src/app/font/actions/useLoadFontFiles.js [hunk 14]
```diff
@@ -37,7 +37,10 @@ function useLoadFontFiles() {
     const handle = cleanForSlug(family);
     const elementId = `web-stories-${handle}-font-css`;
 
-    const hasFontLink = () => document.getElementById(elementId);
+    const hasFontLink = () =>
```
Comment (swissspidy) [inline packages/story-editor/src/app/font/actions/useLoadFontFiles.js]: Is this change needed? Since this is within a callback and never called directly during render as far as I can see

File: packages/story-editor/src/app/story/effects/useHashState.js [hunk 15]
```diff
@@ -34,7 +19,9 @@
 import { useEffect, useState } from '@googleforcreators/react';
 
 export function hashToParams(hash) {
-  return new URLSearchParams(hash.startsWith('#') ? hash.substr(1) : hash);
+  return new URLSearchParams(
+    hash && hash.startsWith('#') ? hash.substr(1) : hash
+  );
```
Comment (swissspidy) [inline packages/story-editor/src/app/story/effects/useHashState.js]:
```suggestion
  return new URLSearchParams(hash?.startsWith('#') ? hash.substr(1) : hash);
```

File: packages/story-editor/src/utils/getBlurHashFromImage.js [hunk 16]
```diff
@@ -26,7 +26,8 @@ import Worker from './generateBlurhash.worker';
 
 const getImageData = (image) => {
   const { width, height } = image;
-  const canvas = document.createElement('canvas');
+  const canvas =
```
Comment (swissspidy) [inline packages/story-editor/src/utils/getBlurHashFromImage.js]: Is this change needed / is this function ever run during SSR?

File: packages/story-editor/src/utils/keyboardOnlyOutline.js [hunk 17]
```diff
@@ -74,7 +74,9 @@ const KeyboardOnlyOutline = () => {
 };
 
 export function isKeyboardUser() {
```
Comment (swissspidy) [inline packages/story-editor/src/utils/keyboardOnlyOutline.js]: This function is only ever called in an effect. Do we still need this?

File: packages/dashboard/src/app/router/routerProvider.js (hunk 8 above)
Comment (mohdsayed) [reply packages/dashboard/src/app/router/routerProvider.js:69]: I am thinking, what would be the possible case when `history.current?.location?.search` may be `null`. We are only adding `.?` here to prevent errors during server-side rendering.

File: .eslintrc (hunk 5 above)
Comment (mohdsayed) [reply .eslintrc]: You are right. 
These are all the `@web-stories-wp` packages.

- e2e-tests
- activation-notice
- commander
- e2e-test-utils
- eslint-import-resolver
- glider
- imgareaselect
- jest-amp
- jest-puppeteer-amp
- jest-resolver
- karma-failed-tests-reporter
- karma-puppeteer-client
- karma-puppeteer-launcher
- stories-block
- stories-carousel
- stories-lightbox
- tinymce-button
- widget
- wp-dashboard
- wp-story-editor
- wp-utils

But not all have React as a dependency, so they would not be using `useEffect` anyways. These are the `@web-stories-wp` packages with React dependency, so added those in the `excludedFiles`.

- activation-notice
- jest-amp
- karma-fixture
- react
- stories-block
- test-utils
- tinymce-button
- wp-dashboard
- wp-story-editor
- wp-utils

File: packages/design-system/src/components/keyboard/index.js (hunk 10 above)
Comment (mohdsayed) [reply packages/design-system/src/components/keyboard/index.js]: Yes, `globalThis.document?.documentElement || null` looks better here. At other places, I used `typeof document !== 'undefined'` because it looked more readable to write 

```js
typeof document !== 'undefined' ? document.getElementById(containerId) : null;`
```

as compared to 

```js
globalThis.document?.getElementById  ? document.getElementById(containerId) : null;`
```

Are you referring to any other instance?

File: packages/dom/src/stripHTML.js (hunk 11 above)
Comment (mohdsayed) [reply packages/dom/src/stripHTML.js:20]: I think we do. Probably will have to use some external library/package for it.

File: packages/dashboard/src/app/router/routerProvider.js (hunk 8 above)
Comment (swissspidy) [reply packages/dashboard/src/app/router/routerProvider.js:69]: During server-side rendering it will be `null`, so the initial state value would be `null` here, no?

File: packages/design-system/src/components/keyboard/index.js (hunk 10 above)
Comment (swissspidy) [reply packages/design-system/src/components/keyboard/index.js]: I don't recall all the instances of either style. I just want to make sure we use a consistent style across the code base.

File: packages/dom/src/stripHTML.js (hunk 11 above)
Comment (mohdsayed) [reply packages/dom/src/stripHTML.js:20]: There are packages available like [himalaya](https://www.npmjs.com/package/himalaya) ( 375kB ) but not very lightweight.

File: packages/i18n/src/translateWithMarkup.js (hunk 13 above)
Comment (mohdsayed) [reply packages/i18n/src/translateWithMarkup.js]: It wouldn't break but return empty strings in the node env. 
There are alternatives like https://www.npmjs.com/package/node-html-parser, but that's 149kB

File: packages/dom/src/stripHTML.js (hunk 11 above)
Comment (swissspidy) [reply packages/dom/src/stripHTML.js:20]: Wouldn't it need something more like [string-strip-html](https://www.npmjs.com/package/string-strip-html)?

Still quite large though, so as long as we don't need it, we can postpone this.

File: packages/dashboard/src/app/router/routerProvider.js (hunk 8 above)
Comment (mohdsayed) [reply packages/dashboard/src/app/router/routerProvider.js:69]: In node env, the initial state would be `undefined` therefore it will return an empty object `{}`. Am I missing something? Will add test coverage with `@jest-environment node`

Comment (swissspidy) [reply packages/dashboard/src/app/router/routerProvider.js:69]: Ah indeed, if it's `undefined` and not `null` then it should be fine.

Comment (mohdsayed) [reply packages/dashboard/src/app/router/routerProvider.js:69]: Was trying to add test coverage for `@jest-environment node` however it doesn't look straightforward as it wouldn't run the same way as it does in SSR. When running a test in a node environment it will also run code in `useEffect`.

Though [React Hooks Testing Library](https://react-hooks-testing-library.com/) does provide [SSR support](https://react-hooks-testing-library.com/usage/ssr) for `renderHook`, however that doesn't seem to work as expected. Let me see if I can find a solution.

File: packages/design-system/src/components/keyboard/index.js (hunk 10 above)
Comment (mohdsayed) [reply packages/design-system/src/components/keyboard/index.js]: Updated all instances in https://github.com/GoogleForCreators/web-stories-wp/pull/11270/commits/438d0fa0537d351bae387dcbf77d54938f9dbdb8

File: packages/dashboard/src/app/router/routerProvider.js (hunk 8 above)
Comment (mohdsayed) [reply packages/dashboard/src/app/router/routerProvider.js:69]: Didn't find any solution for adding test coverage for node env! The doc suggests calling `hydrate()` after `renderHook()`, but that throws `document is not defined` error.

File: packages/dom/src/escapeHTML.js [hunk 18]
```diff
@@ -14,7 +14,7 @@
  * limitations under the License.
  */
 
-const buffer = document.createElement('div');
+const buffer = globalThis?.document?.createElement('div') || {};
```
Comment (swissspidy) [inline packages/dom/src/escapeHTML.js]:
```suggestion
const buffer = globalThis?.document?.createElement('div') || {}; // To handle SSR.
```

File: packages/dom/src/stripHTML.js [hunk 19]
```diff
@@ -14,8 +14,9 @@
  * limitations under the License.
  */
 
-const buffer = document.createElement('div');
+const buffer = globalThis?.document?.createElement('div') || {};
```
Comment (swissspidy) [inline packages/dom/src/stripHTML.js]:
```suggestion
const buffer = globalThis?.document?.createElement('div') || {}; // To handle SSR.
```

File: packages/story-editor/src/app/canvas/canvasProvider.js [hunk 20]
```diff
@@ -62,18 +62,20 @@ function CanvasProvider({ children }) {
...
-          }
-          setBoundingBoxes((boxes) => ({
-            ...boxes,
-            [entry.target.dataset[RECT_OBSERVATION_KEY]]:
-              entry.boundingClientRect,
-          }));
-        }
-      }),
+      globalThis.window
+        ? new window.IntersectionObserver((entries) => {
```
Comment (swissspidy) [inline packages/story-editor/src/app/canvas/canvasProvider.js]: Shouldn't this be like this?

```suggestion
      globalThis.IntersectionObserver
        ? new globalThis.IntersectionObserver((entries) => {
```

**Comment by santoshgistto** [comment]
Looks like  test coverage dropped :)

**Comment by santoshgistto** [comment]
Looks like needs to increase code coverage still 
@sayedtaqui  any help here :)

**Comment by swissspidy** [comment]
Closing for now, best to review this with a fresh PR at some point.
